keps/sig-storage/20191125-bucket-provisioning.md

```diff
@@ -0,0 +1,428 @@
+1. `bucketClassName`: Name of the target BucketClass.
+1. `secretName`: (optional) Secret in the BucketRequest's namespace storing credentials to be used by a workload for bucket access.
+1. `bucketName`: Name of the Bucket instance representing the desired storage instance endpoint.  Set by the COSI Controller for new Buckets.  Specified by the user when connecting to existing buckets.
+1. `phase`: 
+   - *Pending*: The controller has detected the new `Bucket` and begun provisioning operations
+   - *Bound*: Provisioning operations have completed and the `Bucket` has been bound to a `Bucket`.
+
+#### Bucket
+
+A cluster-scoped resource representing a storage instance.  A Bucket should contain enough information to enable the requesting client to access the storage instance.  At a minimum, a Bucket must store enough identifying information such in order for drivers to accurately target the storage instance (e.g. during a deletion process).
+
+```yaml
+apiVersion: cosi.io/v1alpha1
+kind: Bucket
+Metadata:
+  name: [1]
+  labels:
+    cosi.io/provisioner: [2]
+  finalizers:
+  - cosi.io/finalizer [3]
+spec:
+  provisioner: [4]
+  releasePolicy: [5]
+  anonymousAccessMode: [6]
+    - private
+    - publicRead
+    - publicReadWrite
+  bucketClassName: [7]
+  permittedNamespaces: [8]
+    - name:
+      uid:
+  protocol: [9]
+    protocolSignature: ""
+    azureBlob: [10]
+      containerName:
+      storageAccount:
+    s3: [11]
+      endpoint:
+      bucketName:
+      region:
+      signatureVersion:
+    gcs: [12]
+      bucketName:
+      privateKeyName:
+      projectId:
+      serviceAccount:
+  parameters: [13]
+status:
+  message: [14]
+  phase: [15]
+  conditions:
+```
+
+1. `name`: For Greenfield, generated in the pattern of `bucket-<my-bucket>-<my-bucket-namespace>`. 
+1. `labels`: should be added by controller..  Key’s value should be the provisioner name. Characters that do not adhere to [Kubernetes label conventions](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set) will be converted to ‘-’.
+1. `finalizers`: should be added by controller to defer `Bucket` deletion until backend deletion ops succeed.
+1. `provisioner`: The provisioner field defined in the `BucketClass`.  Used by sidecars to filter Buckets.
+1. `releasePolicy`: Prescribes outcome of a Delete events. **Note:** In Brownfield cases, backing storage instance is Retained, and never deleted.
+   - _Delete_:  the bucket and its contents are destroyed.
+   - _Retain_:  the bucket and its data are preserved with only abstracting Kubernetes being destroyed.
+1. `anonymousAccessMode`:  ACL specifying *uncredentialed* access to the Bucket.  This is applicable for cases where the storage instance or objects are intended to be publicly readable and/or writable.  If `BucketClass.anonymousAccessMode` is set, the value is copied here.  Accepted values:
```

Comment:
slight wordsmithing...

Reply:
Need to be clear if these values pertain to the objects within the bucket, or only to the bucket itself. Eg. R can mean user can read objects contained in bucket, or it could mean user can read the bucket (eg. to get a list of its objects).

Reply:
Good question. There is nuanced access modes in object storage systems. ListBucket and getObjectis a different operations, and hence require different permissions. In the case of anonymousAccessMode, we are talking about a more general access pattern, similar to what is seen in file systems. i.e. In case of anonymousAccessMode 'R_ONLY', we allow both ListBucket and GetObject. In case of 'RW', we allow ListBucket,GetObject and PutObject and in case of W_ONLY, we only allow putObject.